src/accuracy/analytical_variance_for_sampling_liquid_legions.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Analytical variance formulas for sampling + LiquidLegions."""
+
+import numpy as np
+from numpy import exp as exp
+from scipy.special import expi as ei
+
+
+class AnalyticalAccuracyEvaluator:
+
+    """Analytical formulas to quickly evaluate the accuracy of sampling + LiquidLegions.
+
+    To estimate reach and frequency histogram using sampling + LiquidLegions, we
+        1. ask each publisher to sample a fraction of vids using the same hash.
+        2. ask each publisher to obtain the the LiquidLegions of the sampled, reached vids, with the same
+            decay rate and number of registers
+        3. Use MPC + SameKeyAggregator protocol to estimate reach and relative frequency histogram of the
+            sampled ids.  Scale up the reach estimate.  Use the relative frequency histogram of the sample
+            as the estimate for population.
+    This class computes the standard deviations of the estimates of
+        - reach, and
+        - proportion of each frequency bucket.
+    TODO(jiayu): derive the stds for k+ reaches and shuffle distance.
+
+    Please read Section of 1 the companion doc:
+    https://drive.google.com/file/d/1FNeMZnSYNMcsAMtzFY2ZyM_Tl1YYoEid/view?usp=sharing&resourcekey=0-NTSWpzbL2HkumqI-AUZlIg
+    for precise problem description and formulas.
+    """
+
+    def __init__(self, a, m, pi, ssreach, ssfreq):
+        """Specify parameters that determine the accuracy.
+
+        Args:
+            a: decay rate of LiquidLegions.
+            m: number of registers in LiquidLegions.
+            pi: sampling fraction, which equals 1 / <number of sampling buckets>.
+            ssreach: variance of noise added on the number of non-empty registers.
+            ssfreq: variance of noise added on each bucket of the frequency histogram of active registers.
+        """
+        self.a = a
+        self.m = int(m)
+        self.pi = pi
+        self.ssreach = ssreach
+        self.ssfreq = ssfreq
+        self.c = a / m / (1 - exp(-a))
+
+    def _local_multiply(self, n, multiplier):
+        """An operation repeatedly used in the formulas."""
```

Comment:
This doesn't seem to be a very useful docstring. Is there a nice interpretation of this quantity? If so, then it might be worthwhile to spell that out instead. A similar comment applies to other _local_* methods below too.

Reply:
+1